executor/grace_hash_join.go

```diff
@@ -0,0 +1,275 @@
+	buildKeyColIdx := make([]int, len(e.buildKeys))
+	for i := range e.buildKeys {
+		buildKeyColIdx[i] = e.buildKeys[i].Index
+	}
+	buildHashCtx := &hashContext{
+		allTypes:  e.buildTypes,
+		keyColIdx: buildKeyColIdx,
+	}
+	e.buildHashPartitioner = newHashPartitioner(e.ctx, e.partitionCnt, buildHashCtx)
+
+	probeKeyColIdx := make([]int, len(e.probeKeys))
+	for i := range e.probeKeys {
+		probeKeyColIdx[i] = e.probeKeys[i].Index
+	}
+	probeHashCtx := &hashContext{
+		allTypes:  e.probeTypes,
+		keyColIdx: probeKeyColIdx,
+	}
+	e.probeHashPartitioner = newHashPartitioner(e.ctx, e.partitionCnt, probeHashCtx)
+
+	return nil
+}
+
+// Close implements the Executor Close interface.
+func (e *GraceHashJoinExec) Close() error {
+	err := e.buildPartitionerExec.partitionInDisk.Close()
+	if err != nil {
+		return err
+	}
+	err = e.probePartitionerExec.partitionInDisk.Close()
+	if err != nil {
+		return err
+	}
+
+	e.buildHashPartitioner = nil
+	e.probeHashPartitioner = nil
+
+	err = e.hashJoinExec.Close()
+	if err != nil {
+		return err
+	}
+
+	err = e.baseExecutor.Close()
+	return err
+}
+
+// partitioning the chunk into partitions and write them to disk.
+func (e *GraceHashJoinExec) partition(chk *chunk.Chunk, hashPartitioner *hashPartitioner, partitionInDisk *chunk.PartitionInDisk) (err error) {
+	if chk.NumRows() == 0 {
+		err := partitionInDisk.Flush()
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	partitions, err := hashPartitioner.split(chk)
+	if err != nil {
+		return err
+	}
+	err = partitionInDisk.Add(chk, partitions)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *GraceHashJoinExec) buildPartitionInDisk(ctx context.Context) (err error) {
+
+	for {
+		chk := chunk.NewChunkWithCapacity(e.buildSideExec.base().retFieldTypes, e.ctx.GetSessionVars().MaxChunkSize)
+		err = Next(ctx, e.buildSideExec, chk)
+		if err != nil {
+			return err
+		}
+		err := e.partition(chk, e.buildHashPartitioner, e.buildPartitionerExec.partitionInDisk)
+		if err != nil {
+			return err
+		}
+		if chk.NumRows() == 0 {
+			break
+		}
+	}
+
+	for {
+		chk := chunk.NewChunkWithCapacity(e.probeSideExec.base().retFieldTypes, e.ctx.GetSessionVars().MaxChunkSize)
+		err = Next(ctx, e.probeSideExec, chk)
+		if err != nil {
+			return err
+		}
+		err := e.partition(chk, e.probeHashPartitioner, e.probePartitionerExec.partitionInDisk)
+		if err != nil {
+			return err
+		}
+		if chk.NumRows() == 0 {
+			break
+		}
+	}
+	return nil
+}
+
+// Next implements the Executor Next interface.
+// grace hash join is executed in two phases:
+// phases 1 is partitioning. fetch data and partitioning them, then write to disk.
+// phases 2 is joining. fetch data from disk and execting join by partition order.
+func (e *GraceHashJoinExec) Next(ctx context.Context, req *chunk.Chunk) (err error) {
+	if !e.prepared {
+		err := e.buildPartitionInDisk(ctx)
+		if err != nil {
+			return err
+		}
+		e.prepared = true
+	}
+
+	req.Reset()
+	if e.partitionIdx == e.partitionCnt {
```

Comment:
There are two questions. 
1. When deciding the number of partitions?
2. In my opinion, this statement is only executed in the case of `e.partitionCnt == 0`. In this case, do we need to generate a warning or report an error?

Reply:
1. I think we can decide the number of partitions in `planner`, based on the data size. I plan to do this in a separate PR
2. In `GraceHashJoinExec`, `partitionCnt` shouldn't be equal to 0, because it means how many pieces we divide chunks. If data size is not too big, we can use `HashJoinExec` directly without partition.
`partitionIdx` is used to indicate which partition is currently being processed. `if e.partitionIdx == e.partitionCnt` check whether we have processed all partitions. If true, we can finish this.